File: docs/trace_template/python/usage.py
"""
Add trace around the real module call. The traced call still returns the original
result, while trace() records module timing and selected output activations.

```diff
  activation("embedding/token_ids.safetensors", token_ids.to(dtype=torch.int64, device="cpu"))

- x = time_mixer(x)
+ x = trace(
+     "cells/cell_0000/time_mixer",
+     time_mixer,
+     x,
+     outputs="cells/cell_0000/time_mixer/embedded_context.safetensors",
+ )

  # This is skipped automatically if x was already saved by the previous module.
  activation("cells/cell_0001/input.safetensors", x)

- loss, lse, max_vals, argmax = L2WRAP_CE_CUDA_V2.forward(logits, targets)
+ loss, lse, max_vals, argmax = trace(
+     "loss/l2wrap_cross_entropy",
+     L2WRAP_CE_CUDA_V2.forward,
+     logits,
+     targets,
+     outputs={
+         0: "loss/l2wrap_cross_entropy.safetensors",
+         1: "loss/l2wrap_cross_entropy/lse.safetensors",
+         2: "loss/l2wrap_cross_entropy/max_vals.safetensors",
+         3: "loss/l2wrap_cross_entropy/argmax.safetensors",
+     },
+ )
```
"""

File: docs/trace_template/cpp/usage.cpp
/*
Add trace around the real module call. The traced call still returns the original
result, while trace_libtorch()/trace_ggml() records module timing and selected
output activations.

```diff
  TraceWriter writer;

- auto logits = head(hidden);
+ auto logits = trace_libtorch(
+     writer,
+     case_root,
+     "lm_head",
+     [&] {
+         return head(hidden);
+     },
+     outputs(out<0>("lm_head/logits.safetensors")),
+     hidden
+ );

  // For ggml / llama.cpp scheduler callbacks, keep the scheduler ask=true/ask=false
  // boundary as the timing source, then save activation and module timing together.
- activation_ggml(writer, case_root, "lm_head/logits.safetensors", logits_node);
+ trace_ggml(
+     writer,
+     case_root,
+     "lm_head",
+     lm_head_elapsed_ns,
+     outputs(node("lm_head/logits.safetensors", logits_node))
+ );
```
*/
